lib/src/game/viewport.dart

```diff
@@ -0,0 +1,185 @@
+/// platforms and players, you should use a viewport.
+///
+/// When using a viewport, [resize] should be called by the engine with
+/// the raw size (on startup and subsequent resizes) and that will configure
+/// [getEffectiveSize()] and [getRawSize()]. The Viewport can also apply
+/// an offset to render and add black (or other color) borders when
+/// necessary. When rendering, call [render] and put all your rendering inside
+/// the lambda so that the correct transformations are applied.
+///
+/// You can think of a Viewport as mechanism to watch a wide-screen movie on a
+/// square monitor. You can stretch the movie to fill the square, but the width
+/// and height will be stretched by different amounts, causing distortion. You
+/// can fill in the smallest dimension and crop the biggest (that causes
+/// cropping). Or you can fill in the biggest and add black bars to cover the
+/// unused space on the smallest (this is the [FixedRatioViewport]).
+///
+/// The other option is to not use a viewport ([DefaultViewport]) and have
+/// your game dynamically render itself to fill in the existing space (basically
+/// this means generating either a wide-screen or a square movie on the fly).
+/// The disadvantage is that different players on different devices will see
+/// different games. For example a hidden door because it's too far away to
+/// render in Screen 1 might be visible on Screen 2. Specially if it's an
+/// online/competitive game, it can give unfair advantages to users with certain
+/// screen resolutions. If you want to "play director" and know exactly what
+/// every player is seeing at every time, you should use a Viewport.
+abstract class Viewport {
+  /// This configures the viewport with a new raw size.
+  /// It should immediately affect [getEffectiveSize] and [getRawSize].
+  /// This must be called by the engine at startup and also whenever the
+  /// size changes.
+  void resize(Vector2 newRawSize);
+
+  /// This transforms the canvas so that the coordinate system is viewport-
+  /// -aware. All your rendering logic should be put inside the lambda.
+  void render(Canvas c, void Function(Canvas c) renderGame);
+
+  /// This returns the effective size, after viewport transformation.
+  /// This is not the game widget size but for all intents and purposes,
+  /// inside your game, this size should be used as the real one.
+  Vector2 getEffectiveSize();
+
+  /// This returns the real widget size (well actually the logical Flutter
+  /// size of your widget). This is the raw size as it would be without any
+  /// viewport.
+  ///
+  /// You probably don't need to care about this if you are using a viewport.
+  Vector2 getRawSize();
+}
+
+/// This is the default viewport if you want no transformation.
+/// The rawSize is just propagated to the effective size and no
+/// translation is applied.
+/// This basically no-ops the viewport.
+class DefaultViewport extends Viewport {
```

Comment:
What about `FullCanvasViewport` or `FullsizeViewport`?

Reply:
I like it being called default because it's super clear it doesn't do anything. FullSize or something could lead someone to believe it will expand or something.